disnake/i18n.py

```diff
@@ -0,0 +1,442 @@
+        A mapping of locales to localized values.
+        Incompatible with ``key``.
+    """
+
+    __slots__ = ("string", "localizations")
+
+    @overload
+    def __init__(self: Localized[StringT], string: StringT, *, key: str):
+        ...
+
+    @overload
+    def __init__(self: Localized[Optional[str]], *, key: str):
+        ...
+
+    @overload
+    def __init__(
+        self: Localized[StringT],
+        string: StringT,
+        *,
+        data: Union[Optional[LocalizationsDict], LocalizationValue],
+    ):
+        ...
+
+    @overload
+    def __init__(
+        self: Localized[Optional[str]],
+        *,
+        data: Union[Optional[LocalizationsDict], LocalizationValue],
+    ):
+        ...
+
+    # note: `data` accepting `LocalizationValue` is intentionally undocumented,
+    # as it's only meant to be used internally
+    def __init__(
+        self,
+        string: StringT = None,
+        *,
+        key: str = MISSING,
+        data: Union[Optional[LocalizationsDict], LocalizationValue] = MISSING,
+    ):
+        self.string: StringT = string
+
+        if not (key is MISSING) ^ (data is MISSING):
+            raise TypeError("Exactly one of `key` or `data` must be provided")
+        if isinstance(data, LocalizationValue):
+            self.localizations = data
+        else:
+            self.localizations = LocalizationValue(key if key is not MISSING else data)
+
+    @overload
+    @classmethod
+    def _cast(cls, string: LocalizedOptional, required: Literal[False]) -> Localized[Optional[str]]:
+        ...
+
+    @overload
+    @classmethod
+    def _cast(cls, string: LocalizedRequired, required: Literal[True]) -> Localized[str]:
+        ...
+
+    @classmethod
+    def _cast(cls, string: Union[Optional[str], Localized[Any]], required: bool) -> Localized[Any]:
+        if not isinstance(string, Localized):
+            string = cls(string, data=None)
+
+        # enforce the `StringT` type at runtime
+        if required and string.string is None:
+            raise ValueError("`string` parameter must be provided")
+        return string
+
+    @overload
+    def _upgrade(self, *, key: Optional[str]) -> Self:
+        ...
+
+    @overload
+    def _upgrade(self, string: str, *, key: Optional[str] = None) -> Localized[str]:
+        ...
+
+    def _upgrade(
+        self, string: Optional[str] = None, *, key: Optional[str] = None
+    ) -> Localized[Any]:
+        # update key if provided and not already set
+        self.localizations._upgrade(key)
+
+        # Only overwrite if not already set (`Localized()` parameter value takes precedence over function names etc.)
+        # Note: not checking whether `string` is an empty string, to keep generic typing correct
+        if not self.string and string is not None:
+            self.string = string
+
+        # this is safe, see above
+        return self
+
+
+class LocalizationValue:
+    """
+    Internal container type for (pending) localization data
```

Comment:
If this is internal, why is it documented?

Reply:
Why is it not meant to be created by the user? Couldn't they use this instead of Localized if they desired?

Reply:
No, this is only used to manage/store the localization key/data, while `Localized` mostly serves as an ephemeral object that gets split up into a string value and a `LocalizationValue` pretty much immediately in most cases, for example:
https://github.com/DisnakeDev/disnake/blob/6d323760a3d93e09a8f6727e6e83705e05ef35a0/disnake/app_commands.py#L135-L137
I'll remove the "Internal" though, since it doesn't really make sense given that this type is part of the documentation

Reply:
https://github.com/DisnakeDev/disnake/pull/269/commits/abcb75c5e6af34d92cf9ab431f48908a875f3e5f